### PR TITLE
Add CreateTokenParser public API

### PR DESCRIPTION
**Pull request by 333fred**
This adds a new public API for creating a token parser that can be repeatedly used on a `SourceText`, with the ability to skip forward and reset back. Closes https://github.com/dotnet/roslyn/issues/72518.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Syntax/SourceTextTokenParser.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            throw new ArgumentOutOfRangeException(nameof(position));
+
+        _lexer.TextWindow.Reset(position);
+    }
+
+    /// <summary>
+    /// Resets the token parser to an earlier position in the input. The parser is reset the start of the token that was previously
+    /// parsed, before any leading trivia, with the directive state that existed at the start of the token.
+    /// </summary>
+    public void ResetTo(Result context)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/SourceTextTokenParser.cs:70]:
```suggestion
    public void ResetTo(Result result)
```

?

'context' is odd here imo.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs [hunk 2]
```diff
@@ -1672,6 +1672,11 @@ public static IEnumerable<SyntaxToken> ParseTokens(string text, int offset = 0,
             }
         }
 
+        public static SyntaxTokenParser CreateTokenParser(SourceText sourceText, CSharpParseOptions? options)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:1675]: i have mixed feelings here.  i would prefer this align with SyntaxFactory.ParseTokens.  That means making options optional.  

i'm on the fence too about taking in an initialTokenPosition.  but i can live with leaving that out.

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/SyntaxTokenParserTests.cs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+    [Fact]
+    public void SkipForwardTo2()
+    {
+        var sourceText = SourceText.From("""class""");
+        var parser = SyntaxFactory.CreateTokenParser(sourceText, TestOptions.Regular);
+        parser.SkipForwardTo(1);
+
+        AssertToken(expectedKind: SyntaxKind.IdentifierToken, expectedContextualKind: SyntaxKind.None, """lass""", parser.ParseNextToken());
+    }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/SyntaxTokenParserTests.cs:143]: test skipping to the end and getting EOF.
test getting multiple EOFs at the end of hte file.
test skipping backwards.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
All nits.

**Review by DustinCampbell — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:1675]: I initially left options optional, but that actually caused a bug in my prototype implementation on the razor side. I'm not hugely opposed to making it optional (as I have now revealed the bug), but that's why I changed it.

I don't think an initial position is necessary. We have a skip method, if you need to change where you start, the tools are there.

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/SyntaxTokenParserTests.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Syntax/LexicalAndXml/SyntaxTokenParserTests.cs:143]: > test getting multiple EOFs at the end of hte file.

Already part of `ParseNextToken` and `DirectiveContextIsPreservedAcrossParseNextToken`.

> test doc comment as leading trivia.

What about that specifically?

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:1675]: so, not pushing on this, but the general idea here is:

1. i as a caller have a chunk of text.   (but which of course then starts at pos 0)
2. that chunk of text represents a location in some larger text construct.
3. when i want to lex tokens in that subtext, i want to their positions to be accurate in the larger text.

but, given teh ability to skip, maybe this is ok.  It's also something we can wait on till the future.

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/SyntaxTokenParserTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/LexicalAndXml/SyntaxTokenParserTests.cs:143]: just have a test that shows it properly lexes, and that you get a DocComment trivia back that you can dive into :)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:1675]: > that chunk of text represents a location in some larger text construct.

Again, this _isn't_ a goal of this API. This method takes a SourceText very specifically. The goal of this API is that you can parse C# tokens from a given source text, with accurate positions inside that source text. If you want to do some futher mapping where you created a source text from a smaller chunk of the original, that's fine, but that additional mapping is on the consumer of the API.

**Comment by CyrusNajmabadi** [comment]
So one of hte ways i operate is to think about how we might ensure consistency between the apis we have already.  In a very real sense, i think it would make sense for `SyntaxFactory.ParseTokens` to just become a think wrapper around this SyntaxTokenParser.  That way we have uniform and consistent behavior between the two, just with the new API providing a superset of capabilities (working on source-texts, and supporting jumping forward/back).

I get that you don't have that as a goal here.  But imo the API is cleaner and healthier if we don't think of these as two entirely independent paths.  But rather an existing limited path, and the new path that legacy path sits on which supports a superset of behavior.  Then, instead of having two entirely different entrypoints with varying different support.  We just have the simple/legacy entrypoint, and the modern/rich entrypoint, and we could have anyone who sits on the old one feel they could move uniformly over tot he new one.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews

**Review by chsienki — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Comment by jaredpar** [comment]
@cston PTAL

File: src/Compilers/CSharp/Portable/Syntax/SourceTextTokenParser.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using InternalSyntax = Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+/// <summary>
+/// A token parser that can be used parse tokens continuously from a source. This parser parses continuously; every call to
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/SourceTextTokenParser.cs:12]: >used parse [](http://example.com/codeflow?start=31&length=10)

"used *to* parse"?

File: src/Compilers/CSharp/Portable/Syntax/SourceTextTokenParser.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    public void SkipForwardTo(int position)
+    {
+        if (position < _lexer.TextWindow.Position)
+            throw new ArgumentOutOfRangeException(nameof(position));
+
+        _lexer.TextWindow.Reset(position);
+    }
+
+    /// <summary>
+    /// Resets the token parser to an earlier position in the input. The parser is reset the start of the token that was previously
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/SourceTextTokenParser.cs:71]: >reset the start [](http://example.com/codeflow?start=83&length=15)

"reset *to* the start"?

**Review by cston — APPROVED** [review]

**Comment by RikkiGibson** [comment]
into 17.10 it goes!

**Pull request merged**